<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-надежности-заемщиков." data-toc-modified-id="Исследование-надежности-заемщиков.-0"><span class="toc-item-num">0&nbsp;&nbsp;</span><strong>Исследование надежности заемщиков.</strong></a></span></li><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.0.4"><span class="toc-item-num">2.0.4&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-2.0.5"><span class="toc-item-num">2.0.5&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-2.0.6"><span class="toc-item-num">2.0.6&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li></ul></li></ul></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Влияет-ли-семейное-положение-и-количество-детей-на-факт-погашения-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Влияет-ли-семейное-положение-и-количество-детей-на-факт-погашения-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вопрос 1: Влияет ли семейное положение и количество детей на факт погашения кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Ответ-1.--В-результате-анализа-установлено,-что-:" data-toc-modified-id="Ответ-1.--В-результате-анализа-установлено,-что-:-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span><strong>Ответ 1.  В результате анализа установлено, что :</strong></a></span></li></ul></li><li><span><a href="#Вопрос-2:-Влияет-ли-уровень-дохода-и-цель-кредита-на-погашения-кредита-в-срок?¶" data-toc-modified-id="Вопрос-2:-Влияет-ли-уровень-дохода-и-цель-кредита-на-погашения-кредита-в-срок?¶-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вопрос 2: Влияет ли уровень дохода и цель кредита на погашения кредита в срок?¶</a></span><ul class="toc-item"><li><span><a href="#Ответ-2.-В-результате-анализа-установлено,-что-:" data-toc-modified-id="Ответ-2.-В-результате-анализа-установлено,-что-:-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span><strong>Ответ 2. В результате анализа установлено, что :</strong></a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

**Исследование надежности заемщиков.**
=====================================

# Шаг 1. Обзор данных

In [1]:
# импортируем библиотеку pandas
import pandas as pd 
#сохраним таблицу в переменной client_stat
client_stat = pd.read_csv('/datasets/data.csv') 
#познакомимся со структурой таблицы client_stat начало
display(client_stat.head(10)) 



children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [2]:
#познакомимся со структурой таблицы client_stat конец
display(client_stat.tail(10)) 


children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

In [3]:
#получим общую информацию о датасете client_stat
client_stat.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Результат обзора.** 
В таблице 12 колонок и 21525 строк. В столбцах days_employed и total_income есть пропущенные значения, которые следует проанализировать и заменить на другие.

# Шаг 2.1 Заполнение пропусков

In [4]:
#поиск всех уникальных значений в стролбце
display('Уникальные значения в колонке с общим трудовым стажем:',client_stat['days_employed'].unique()) 
#количество пропусков
print('Количество пропусков в колонке с общим трудовым стажем:',client_stat['days_employed'].isna().sum()) 


'Уникальные значения в колонке с общим трудовым стажем:'

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

Количество пропусков в колонке с общим трудовым стажем: 2174


**Вывод:** Отрицательные значения с колонке с общим трудовым стажем могут появиться по причине ошибочно заполненной формы: данные начала и конца трудовой деятельности перепунаны местами или не заполнена одна из  ячеек. Поэтому отрицательные значения для анализа сохраним как положительные методом abs()

In [5]:
#избавимся от отрицательных значений
client_stat['days_employed'] = client_stat['days_employed'].abs()


In [6]:
#отсортируем  по убыванию колонки'days_employed' 
display(client_stat.sort_values('days_employed',ascending=False))


children  days_employed  dob_years education  education_id  \
6954          0  401755.400475         56   среднее             1   
10006         0  401715.811749         69    высшее             0   
7664          1  401675.093434         61   среднее             1   
2156          0  401674.466633         60   среднее             1   
7794          0  401663.850046         61   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender income_type  debt  \
6954          вдовец / вдова                 2      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7794        гражданский брак                 1      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21489        женат / замужем                 0      M   компаньон     0   
21495       гражданский брак                 1      F   сотрудник     0   
21497        женат / замужем                 0      F   компаньон     0   
21502        женат / замужем                 0      F   сотрудник     0   
21510        женат / замужем                 0      F   сотрудник     0   

        total_income                           purpose  
6954   176278.441171                      ремонт жилью  
10006   57390.256908             получение образования  
7664   126214.519212                 операции с жильем  
2156   325395.724541                        автомобили  
7794    48286.441362                           свадьба  
...              ...                               ...  
21489            NaN              сделка с автомобилем  
21495            NaN                           свадьба  
21497            NaN        строительство недвижимости  
21502            NaN  строительство жилой недвижимости  
21510            NaN           приобретение автомобиля  

[21525 rows x 12 columns]

**Вывод:** При анализе показаний стажа, убеждаемся, что максимальное показание стажа соответствует более взрослым клиентам. 

In [7]:
# проверим сколько значений более 14000 в client_stat
display(client_stat[client_stat['days_employed']>13260].value_counts())

children  days_employed  dob_years  education  education_id  family_status          family_status_id  gender  income_type  debt  total_income   purpose                              
-1        340499.039342  50         среднее    1             женат / замужем        0                 M       пенсионер    0     170762.773162  покупка своего жилья                     1
 0        381198.918686  58         среднее    1             женат / замужем        0                 F       пенсионер    0     220571.822007  покупка недвижимости                     1
          380893.813588  61         среднее    1             женат / замужем        0                 M       пенсионер    0     533870.927357  покупка жилья                            1
          380899.304883  52         среднее    1             женат / замужем        0                 F       пенсионер    0     161296.851590  строительство недвижимости               1
          380908.859717  53         среднее    1             Не женат 

<div class="alert alert-info">
    
Количество неправдоподобных значений более 10% (3470/21525%=16,12%), поэтому удалять их не будем, а заменим на правдоподобное.
</div>


In [8]:
# в функции total_days_employed примем максимально возможный стаж за 13260 дней и все значения больше 13260 заменим на 13260

def total_days_employed(days_employed):
    if days_employed > 13260:
        days_employed==13260
        return 13260
    
    return days_employed
    
    
print(total_days_employed(10000)) # проверим работу функции для  10000
print(total_days_employed(235000)) # проверим работу функции для  235000



10000
13260


In [9]:
# используя метод apply заменим в таблице значения стажа более 13260 на 13260
client_stat['days_employed'] = client_stat['days_employed'].apply(total_days_employed)

# проконтролируем внесённые изменения
display(client_stat.sort_values('days_employed',ascending=False))

children  days_employed  dob_years education  education_id  \
10714         0        13260.0         63    высшее             0   
2781          0        13260.0         54   среднее             1   
10345         0        13260.0         56   среднее             1   
10340         0        13260.0         56   среднее             1   
10336         0        13260.0         65   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender income_type  debt  \
10714         вдовец / вдова                 2      F   пенсионер     0   
2781         женат / замужем                 0      F   пенсионер     0   
10345       гражданский брак                 1      F   пенсионер     0   
10340  Не женат / не замужем                 4      F   пенсионер     0   
10336         вдовец / вдова                 2      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
21489        женат / замужем                 0      M   компаньон     0   
21495       гражданский брак                 1      F   сотрудник     0   
21497        женат / замужем                 0      F   компаньон     0   
21502        женат / замужем                 0      F   сотрудник     0   
21510        женат / замужем                 0      F   сотрудник     0   

        total_income                           purpose  
10714  112734.356402           покупка жилья для семьи  
2781   210536.558888              покупка недвижимости  
10345  161380.764539        строительство недвижимости  
10340  355525.588501        дополнительное образование  
10336  266426.566475              покупка своего жилья  
...              ...                               ...  
21489            NaN              сделка с автомобилем  
21495            NaN                           свадьба  
21497            NaN        строительство недвижимости  
21502            NaN  строительство жилой недвижимости  
21510            NaN           приобретение автомобиля  

[21525 rows x 12 columns]

<div class="alert alert-info">
    
Результат: изменения неправдоподобно больших значений стажа работы внесены в таблицу client_stat
</div>


In [10]:
# пропуски в колонке days_employed
display(client_stat[client_stat['days_employed'].isna()].head(10)) 


children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

In [11]:
#количество пропусков
print('Количество пропусков в колонке с ежемесячным доходом:',client_stat['total_income'].isna().sum())


Количество пропусков в колонке с ежемесячным доходом: 2174


In [12]:
#объдиним для анализа эти колонки в одну таблицу
client_income_employed = client_stat[['days_employed','total_income']]
#посмотрим какая связь в показаниях  обеих колонок
print(client_income_employed.sort_values('days_employed',ascending=False)) 


       days_employed   total_income
10714        13260.0  112734.356402
2781         13260.0  210536.558888
10345        13260.0  161380.764539
10340        13260.0  355525.588501
10336        13260.0  266426.566475
...              ...            ...
21489            NaN            NaN
21495            NaN            NaN
21497            NaN            NaN
21502            NaN            NaN
21510            NaN            NaN

[21525 rows x 2 columns]


**Вывод:** Пропуски значений в колонках со стажем и ежемесячным доходом одновременно скорее всего связаны с тем, что заемщик может быть безработным или не имеет документального подтверждения стажа и дохода. Точной информации у нас по этому поводу нет. Количество пропущенных значений более 10 %, а также в данных таблицы есть информация, что дети у таких клиентов есть.Следовательно, для цели исследования, игнорировать пропуски не стоит, а имеет смысл заменить медианным значением.Медианное значение предпочтительней среднего арифметического, т.к. "сглаживает" влияние крайних значений .

In [13]:
#найдём минимальные значения для каждой из колонок
print(client_income_employed.min()) 

days_employed       24.141633
total_income     20667.263793
dtype: float64


**Вывод:** *отрицательных значений в колонках days_employed и total_income нет*

In [14]:
#найдём медианные значения для каждой из колонок
print(client_income_employed.median() ) 

days_employed      2194.220567
total_income     145017.937533
dtype: float64


In [15]:
 #заменим пропуски на медианные значения 
client_stat['days_employed'] = client_stat['days_employed'].fillna(client_stat['days_employed'].median()) 
# контроль количества пропусков
print(client_stat['days_employed'].isna().sum()) 
 #контроль по общим данным таблицы- пропущенных значений  в колонках нет
client_stat.info()

0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [16]:
#заменим пропуски на медианные значения  
client_stat['total_income'] = client_stat['total_income'].fillna(client_stat['total_income'].median()) 
#контроль количества пропусков
print(client_stat['total_income'].isna().sum()) 
 #общая информация о таблице
client_stat.info()

0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Итог :** Пропуски в колонках 'days_employed' и 'total_income' заменены на мадианные значения, изменения внесены в таблицу client_stat. 

### Шаг 2.2 Проверка данных на аномалии и исправления.

* ***В колонке 'education'значения указаны в разных регистрах.***

In [17]:
#изменим регистр в 'education'
client_stat['education'] = client_stat['education'].str.lower() 
 #контроль изменений регистра
display(client_stat.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   13260.000000         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

**Результат:** В колонке 'education' привели все строки к одному регистру и сохранили изменения в таблице.

In [18]:
#проверим колонки "children" на аномалии
client_stat_children = client_stat['children'] 
# min и max значения таблицы
print(client_stat_children.agg(['min', 'max'])) 
# ознакомимся с значениями таблицы
print(client_stat_children.value_counts().head(10)) 

min    -1
max    20
Name: children, dtype: int64
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


* ***47 отрицательных значений в колонке 'children' составляют менее 1%, поэтому можно заменить эти значения на 1 методом abs() без ущерба для исспедования.***

In [19]:
# заменим отрицательные значения на положительные
client_stat['children'] = client_stat['children'].abs() 
#контроль изменений 
print(client_stat_children.value_counts().head(10)) 

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64


In [20]:
#исправим значения 20 в колонке 'children'  на 2
client_stat.loc[(client_stat['children']==20),'children']=2 
 # выведем на экран результат исправлений
print(client_stat['children'].value_counts().head(10))

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


**Результат:** В колонке *'children'* отрицательные значения заменены на положительные, количество 20 в 76 случаях заменены на 2, т.к. очень похоже на опечатку.Если бы это были по-настоящему многодетные семьи или семьи с приёмными детьми, значения были бы не такими круглыми(11,9,17 и т.д).

### Шаг 2.3. Изменение типов данных.

In [21]:
# заменим тип данных - вещественные числа на целочисленные показатели
client_stat['total_income'] = client_stat['total_income'].astype('int') 
#контроль изменений
print(client_stat['total_income'].head(10))  


0    253875
1    112080
2    145885
3    267628
4    158616
5    255763
6    240525
7    135823
8     95856
9    144425
Name: total_income, dtype: int64


**Результат:** В колонке *'total_income'* заменили вещественный тип данных на целочисленный.

### Шаг 2.4. Удаление дубликатов.

In [22]:
#проверим данные на наличие и количество дубликатов
print('Количество дубликатов:',client_stat.duplicated().sum()) 


Количество дубликатов: 71


In [23]:
#сохраним дубликаты в отдельную таблицу client_stat_duplicates
client_stat_duplicates = client_stat[client_stat.duplicated()]
#просмотрим первые 10 строк этой таблицы
display(client_stat_duplicates.head(10)) 

children  days_employed  dob_years education  education_id  \
2849         0    2194.220567         41   среднее             1   
3290         0    2194.220567         58   среднее             1   
4182         1    2194.220567         34    высшее             0   
4851         0    2194.220567         60   среднее             1   
5557         0    2194.220567         58   среднее             1   
6312         0    2194.220567         30   среднее             1   
7808         0    2194.220567         57   среднее             1   
7921         0    2194.220567         64    высшее             0   
7938         0    2194.220567         71   среднее             1   
8583         0    2194.220567         58    высшее             0   

              family_status  family_status_id gender income_type  debt  \
2849        женат / замужем                 0      F   сотрудник     0   
3290       гражданский брак                 1      F   пенсионер     0   
4182       гражданский брак                 1      F   сотрудник     0   
4851       гражданский брак                 1      F   пенсионер     0   
5557       гражданский брак                 1      F   пенсионер     0   
6312        женат / замужем                 0      M   сотрудник     0   
7808       гражданский брак                 1      F   пенсионер     0   
7921       гражданский брак                 1      F   пенсионер     0   
7938       гражданский брак                 1      F   пенсионер     0   
8583  Не женат / не замужем                 4      F   пенсионер     0   

      total_income                           purpose  
2849        145017           покупка жилья для семьи  
3290        145017                   сыграть свадьбу  
4182        145017                           свадьба  
4851        145017                           свадьба  
5557        145017                   сыграть свадьбу  
6312        145017  строительство жилой недвижимости  
7808        145017             на проведение свадьбы  
7921        145017             на проведение свадьбы  
7938        145017             на проведение свадьбы  
8583        145017        дополнительное образование

In [24]:
#удаление дубликатов
client_stat = client_stat.drop_duplicates().reset_index(drop=True)
#контроль удаления дубликатов
print('Количество дубликатов:',client_stat.duplicated().sum()) 

Количество дубликатов: 0


**Результат:** Дубликаты удалены методом с восстановлением индексации.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [25]:
# создадим таблицу где каждому значению'education' соответствует уникальное значение 'education_id'
client_stat_education = client_stat[['education','education_id']]
#просмотрим первые 10 строк таблицы 
print(client_stat_education.head(10)) 


  education  education_id
0    высшее             0
1   среднее             1
2   среднее             1
3   среднее             1
4   среднее             1
5    высшее             0
6    высшее             0
7   среднее             1
8    высшее             0
9   среднее             1


In [26]:

print('Количество дубликатов:',client_stat_education.duplicated().sum()) 
#удаление дубликатов
client_stat_education = client_stat_education.drop_duplicates().reset_index(drop=True) 

print('Количество дубликатов:',client_stat_education.duplicated().sum()) 
#просмотрим что получилось
print(client_stat_education) 

Количество дубликатов: 21449
Количество дубликатов: 0
             education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4


In [27]:
# создадим таблицу где каждому значению'family_status' соответствует уникальное значение 'family_status_id' 
client_stat_family = client_stat[['family_status','family_status_id']]
#просмотрим первые 10 строк таблицы
print(client_stat_family.head(10))

      family_status  family_status_id
0   женат / замужем                 0
1   женат / замужем                 0
2   женат / замужем                 0
3   женат / замужем                 0
4  гражданский брак                 1
5  гражданский брак                 1
6   женат / замужем                 0
7   женат / замужем                 0
8  гражданский брак                 1
9   женат / замужем                 0


In [28]:
#удаление дубликатов
print('Количество дубликатов:',client_stat_family.duplicated().sum()) 

client_stat_family = client_stat_family.drop_duplicates().reset_index(drop=True) 

print('Количество дубликатов:',client_stat_family.duplicated().sum()) 
#просмотрим что получилось
display(client_stat_family)

Количество дубликатов: 21449
Количество дубликатов: 0


family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [29]:
# удалены столбцы 'education' и 'family_status' 
client_stat_id=client_stat.drop(columns=['education','family_status'])
# просмотр 10 строк нового датафрейма
display(client_stat_id.head(10)) 


children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   13260.000000         53             1                 1      F   
5         0     926.185831         27             0                 1      M   
6         0    2879.202052         43             0                 0      F   
7         0     152.779569         50             1                 0      M   
8         2    6929.865299         35             0                 1      F   
9         0    2188.756445         41             1                 0      M   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу  
5   компаньон     0        255763               покупка жилья  
6   компаньон     0        240525           операции с жильем  
7   сотрудник     0        135823                 образование  
8   сотрудник     0         95856       на проведение свадьбы  
9   сотрудник     0        144425     покупка жилья для семьи

**Результат:** Для удобства обработки заменили текстовые колонки *education* и *family_status* на целочисленные идентификаторы в колонках *education_id* и *family_status_id* соответственно.

### Шаг 2.6. Категоризация дохода.

In [30]:
# зададим функцию для содания подгрупп
def total_income_category(total_income):
    if total_income <= 30000:
        return 'E'
    if total_income <= 50000 and total_income >=30001:
        return 'D'
    if total_income <= 200000 and total_income >=50001:
        return 'C'
    if total_income <= 1000000 and total_income >=200001:
        return 'B'
    return 'A'
# проверим работу функции для дохода 25000
print(total_income_category(25000)) 
# проверим работу функции для дохода 235000
print(total_income_category(235000)) 

E
B


* **Функция для создания столбца *total_income_category* позволит классифицировать клиентов по размеру дохода.**

In [31]:
# создадим столбец с категориями дохода 'A','B','C','D','E'
client_stat_id['total_income_category'] = client_stat_id['total_income'].apply(total_income_category)
display(client_stat_id.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   13260.000000         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

In [32]:
print(client_stat_id['total_income_category'].value_counts())

C    16016
B     5041
D      350
A       25
E       22
Name: total_income_category, dtype: int64


***Вывод:*** большинство заемщиков имеют доход от 50001 до 200000 рублей, наименьшую долю составляют заёмщики с максимальным( более 1000000 руб.) и минимальным доходом (менее 30000 руб.).

### Шаг 2.7. Категоризация целей кредита.

In [33]:
print(client_stat['purpose'].value_counts()) # выведем на экран все значения  колонки 'purpose'

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

* **Значения колонки 'purpose' имеют множество неявных дубликатов. Чтобы объединить сходные значения, воспользуемся специальной функцией.**

In [34]:
#сохраним колонку 'purpose' в отдельный Series
client_stat_purpose = client_stat[[ 'purpose']] 
# выведем на печать первые 10 значений полученной таблицы
print(client_stat_purpose.head(10))  


                      purpose
0               покупка жилья
1     приобретение автомобиля
2               покупка жилья
3  дополнительное образование
4             сыграть свадьбу
5               покупка жилья
6           операции с жильем
7                 образование
8       на проведение свадьбы
9     покупка жилья для семьи


Функция распределит  все значения колонки 'purpose' по основным группам: 
*  **'получение образования'**
*  **'операции с автомобилем '**
*  **'проведение свадьбы'**
*  **'операции с недвижимостью'**


In [35]:
#функция для группировки по ключевым словам
def purpose_category (purpose): 
            
    if 'образов'in purpose:
        return 'получение образования'
    
    if 'автомоб' in purpose :
        return 'операции с автомобилем '
    
    if 'свадьб' in purpose :
        return 'проведение свадьбы'
       
    return 'операции с недвижимостью'
# проверка работы функции
print(purpose_category('свадьб')) 
# проверка работы функции
print(purpose_category('дополнительное образование')) 
# проверка работы функции
print(purpose_category('покупка жилья для семьи')) 
# проверка работы функции
print(purpose_category('свой автомобиль')) 

проведение свадьбы
получение образования
операции с недвижимостью
операции с автомобилем 


In [36]:
#добавим полученный столбец к таблице *client_stat_purpose* методом *apply* 
client_stat_purpose['purpose_category'] = client_stat_purpose['purpose'].apply(purpose_category)

#выведем на экран первые 10 строк полученной таблицы
display(client_stat_purpose.head(10)) 


/tmp/ipykernel_68/1964235254.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  client_stat_purpose['purpose_category'] = client_stat_purpose['purpose'].apply(purpose_category)


purpose          purpose_category
0               покупка жилья  операции с недвижимостью
1     приобретение автомобиля   операции с автомобилем 
2               покупка жилья  операции с недвижимостью
3  дополнительное образование     получение образования
4             сыграть свадьбу        проведение свадьбы
5               покупка жилья  операции с недвижимостью
6           операции с жильем  операции с недвижимостью
7                 образование     получение образования
8       на проведение свадьбы        проведение свадьбы
9     покупка жилья для семьи  операции с недвижимостью

**Результат:** Категории целевого назначения *'purpose_category'*соответствуют исходным данным колонки *'purpose'*. Теперь можно заменить в исходной таблице значения *'purpose'* на значения *'purpose_category'*. Это должно упростить дальнейший анализ.

In [37]:
#заменим назначение кредита в столбце 'purpose' на значения после группировки по категориям 'purpose_category' 
client_stat['purpose']=client_stat_purpose['purpose_category']
client_stat_id['purpose']=client_stat_purpose['purpose_category']
#сгруппируем полученную таблицу по столбцу и запросим количество кредитных задолженностей
client_stat.groupby('purpose')['debt'].sum()     

purpose
операции с автомобилем      403
операции с недвижимостью    782
получение образования       370
проведение свадьбы          186
Name: debt, dtype: int64

In [38]:
# проанализируем средний показатель задолженности в зависимости от цели кредита
client_stat_id.groupby('purpose').debt.agg(['count','sum','mean'])

count  sum      mean
purpose                                       
операции с автомобилем     4306  403  0.093590
операции с недвижимостью  10811  782  0.072334
получение образования      4013  370  0.092200
проведение свадьбы         2324  186  0.080034

<div class="alert alert-info">
    
    
Анализируя показатель средний задолженности в зависимости от цели кредита установлено, что наиболее рискованными являются кредиты на операциями с автомобилем (0.093590) и на получение образования (0.092200). Более надёжные кредиты выдаются на операции с недвижимостью (0.072334).
 </div>   

In [39]:
#запросим количество кредитов  с нарушениями (индекс 1)
print(client_stat_id['debt'].value_counts()) 


0    19713
1     1741
Name: debt, dtype: int64


In [40]:
 #выведем на экран количество кредитных задолженностей сгруппированные по колонке'children'
print(client_stat_id.groupby('children')['debt'].sum()) 



children
0    1063
1     445
2     202
3      27
4       4
5       0
Name: debt, dtype: int64


<div class="alert alert-info">
    
Доля клиентов с задолженностью:1741/(19713+1741)%=8,12%.
   * Из них не имеют детей: 1063/(19713+1741)%=4,95%
   *  имеют одного ребёнка: 445/(19713+1741)%=2,07%
   *  имеют двоих детей: 202/(19713+1741)%=0,94%
   *  имеют троих детей: 27/(19713+1741)%=0,13%
   *  имеют четверых детей: 4/(19713+1741)%=0,02%
   *  имеют пятерых детей: 0%
</div>


In [41]:
# проанализируем средний показатель задолженности в зависимости от количества детей
client_stat_id.groupby('children').debt.agg(['count','sum','mean'])

count   sum      mean
children                       
0         14091  1063  0.075438
1          4855   445  0.091658
2          2128   202  0.094925
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

<div class="alert alert-info">
    
* Средний показатель задолженности в группе без детей(0.075438) ниже чем в группах с детьми, максимальный показатель- в группе с 4 детьми (0.097561), но она мало показательна, т.к. количество таких семей, как и семей с 3 детьми среди заемщиков невелико. Если сравнивать более многочисленные группы с одним и двумя детьми, то динамика показывает рост показателя с ростом количества детей: с 0.091658 в семьях с одним ребенком до 0.094925 в семьях с двумя детьми. Поэтому можно сделать вывод, что количество детей в семьях заемщиков увеличивает возможность возникновения кредитной задолженности.
    </div>

In [42]:
#сохраним в отдельную переменную данные по кредитам без задолженностей
client_stat_id_debit_0 = client_stat_id[client_stat_id['debt']==0] 

#выведем на экран количество кредитов семьям с детьми
print('Не имели задолженность по возврату кредита и имеют детей:',client_stat_id_debit_0[client_stat_id_debit_0['children']>0]['debt'].count()) 

# сгруппируем полученные данные по столбцу 'children' выведем количество кредитной задолженности
client_children_debit_0 = client_stat_id_debit_0.groupby('children')['debt'].count()

print('Отсутствует задолженность по возврату кредита и количество детей:',client_children_debit_0)


Не имели задолженность по возврату кредита и имеют детей: 6685
Отсутствует задолженность по возврату кредита и количество детей: children
0    13028
1     4410
2     1926
3      303
4       37
5        9
Name: debt, dtype: int64


<div class="alert alert-info">
    
Доля клиентов без задолженности:19713/(19713+1741)%=91,88%.Из них:
    
    *  не имеют дететей: 13028/(19713+1741)%=60,73%
    *  имеют одного ребёнка: 4410/(19713+1741)%=20,56%
    *  имеют двоих детей: 1926/(19713+1741)%=8,98%
    *  имеют троих детей: 303/(19713+1741)%=1,41%
    *  имеют четверых детей: 37/(19713+1741)%=0,17%
    *  имеют пятеро детей:9/(19713+1741)%=0,04%
</div>


In [43]:
# оставим для анализа только данные по  кредитной задолженности (индекс 1)
client_stat_id_debit_1 = client_stat_id[client_stat_id['debt']==1]

#выведем на экран количество кредитных нарушителей с детьми
print('Имели задолженность по возврату кредита и имеют детей:',client_stat_id_debit_1[client_stat_id_debit_1['children']>0]['debt'].sum()) 

# сгруппируем полученные данные по столбцу 'children' 
client_children_debit_1 = client_stat_id_debit_1.groupby('children')['debt'].count()

# выведем количество кредитной задолженности
print('Задолженность по возврату кредита и количество детей:',client_children_debit_1)


Имели задолженность по возврату кредита и имеют детей: 678
Задолженность по возврату кредита и количество детей: children
0    1063
1     445
2     202
3      27
4       4
Name: debt, dtype: int64


<div class="alert alert-info">
    
Доля клиентов с детьми и задолженностью: 678/(19713+1741)%=3,16%. 
    
    Из них
    *  имеют одного ребёнка: 445/(19713+1741)%=2,07%
    *  имеют двоих детей: 202/(19713+1741)%=0,94%
    *  имеют троих детей: 27/(19713+1741)%=0,13%
    *  имеют четверых детей: 4/(19713+1741)%=0,02%
</div>


**Вывод:** Наличие детей у заёмщиков влияет на факт погашения кредита в срок положительно- менее 39% (678 из 1741 )случаев возникает в семьях с детьми. Cемьи с 5 детьми не имели задолженности по кредитам совсем.  Доля кредитов с нарушениями у клиентов с детьми-3,16%(678/(19713+1741)%), у клиентов без детей-4,95%(1063/(19713+1741)%). Разница 1,79 % не значительна.  Внутри группы с нарушениями у клиентов с детьми процент нарушений в зависимости от количетсва детей колеблется тоже незначительно (в пределах 2%- от 0,02 до 2,07).

In [44]:
# сгруппируем таблицу по колонке 'family_status_id'  и значениям  столбца 'debt'
client_stat_id_debit_1.groupby('family_status_id')['debt'].sum()

family_status_id
0    931
1    388
2     63
3     85
4    274
Name: debt, dtype: int64

In [45]:
# проанализируем средний показатель задолженности в зависимости от семейного положения
client_stat_id.groupby('family_status_id').debt.agg(['count','sum','mean'])

count  sum      mean
family_status_id                      
0                 12339  931  0.075452
1                  4151  388  0.093471
2                   959   63  0.065693
3                  1195   85  0.071130
4                  2810  274  0.097509

**Вывод:** 
<div class="alert alert-info">


Анализ данных по кредитной задолженнсти в зависимости от семейного положения показал :

* женатые/замужние клиенты, а также разведённые и овдовевшие, реже нарушают сроки  кредитных договоров, чем другие группы (средний показатель по группе от 0.065693 до 0.075452)
* наиболее высокий показатель в группе одиноких клиентов и клиентов в гражданском браке (от 0.093471 до 0.097509)
    </div>

In [46]:
# сгруппируем таблицу по колонке 'total_income_category'  и значениям  столбца 'debt'
client_stat_id_debit_1.groupby('total_income_category')['debt'].sum()


total_income_category
A       2
B     356
C    1360
D      21
E       2
Name: debt, dtype: int64

In [47]:
# проанализируем средний показатель задолженности в зависимости от уровня дохода
client_stat_id.groupby('total_income_category').debt.agg(['count','sum','mean'])

count   sum      mean
total_income_category                       
A                         25     2  0.080000
B                       5041   356  0.070621
C                      16016  1360  0.084915
D                        350    21  0.060000
E                         22     2  0.090909

<div class="alert alert-info">

    
**Вывод:** 

Наибольшее количество нарушений по возвратам кредита (0.084915) у клиентов с уровнем   дохода  от 50000 руб. до 200000 руб. Еще выше показатель 0.090909 в группе с доходом менее 30000 руб., но она малочислена и поэтому для анализа не подходит, как и группа с доходом более 2000000 руб. с показателем 0.080000 .
Наименьший показатель 0.060000 в группе  клиентов с доходом от 30000 руб. до 50000 руб.
    </div>

# Ответы на вопросы.

## Вопрос 1: Влияет ли семейное положение и количество детей на факт погашения кредита в срок?


### **Ответ 1.  В результате анализа установлено, что :**
<div class="alert alert-info">
   
    
* Средний показатель задолженности в группе без детей(0.075438) ниже чем в группах с детьми, максимальный показатель- в группе с 4 детьми (0.097561), но она мало показательна, т.к. количество таких семей, как и семей с 3 детьми среди заемщиков невелико. Если сравнивать более многочисленные группы с одним и двумя детьми, то динамика показывает рост показателя с ростом количества детей: с 0.091658 в семьях с одним ребенком до 0.094925 в семьях с двумя детьми. Поэтому можно сделать вывод, что количество детей в семьях заемщиков может увеличивает возможность возникновения кредитной задолженности.
    
* Анализ данных по кредитной задолженнсти в зависимости от семейного положения показал :

* женатые/замужние клиенты, а также разведённые и овдовевшие, реже нарушают сроки  кредитных договоров, чем другие группы (средний показатель по группе от 0.065693 до 0.075452)
* наиболее высокий показатель в группе одиноких клиентов и клиентов в гражданском браке (от 0.093471 до 0.097509)
    </div>

##  Вопрос 2: Влияет ли уровень дохода и цель кредита на погашения кредита в срок?¶

###  **Ответ 2. В результате анализа установлено, что :**
<div class="alert alert-info">
    

* Наибольшее количество нарушений по возвратам кредита (0.084915) у клиентов с уровнем дохода от 50000 руб. до 200000 руб. Еще выше показатель 0.090909 в группе с доходом менее 30000 руб., но она малочислена и поэтому для анализа не подходит, как и группа с доходом более 2000000 руб. с показателем 0.080000 . Наименьший показатель 0.060000 в группе клиентов с доходом от 30000 руб. до 50000 руб.
    
* Цель кредита влияет на факт погашения кредита в срок.
Анализируя показатель средний задолженности в зависимости от цели кредита установлено, что наиболее рискованными являются кредиты на операциями с автомобилем (0.093590) и на получение образования (0.092200). Более надёжные кредиты выдаются на операции с недвижимостью (0.072334).
    </div>

# Общий вывод:

<div class="alert alert-info">
    
    
Запрос банка состоял в следующем:***влияет ли семейное положение и количество детей на факт погашения кредита в срок.***
В результате анализа представленных банком статистических данных установлено:
    
* семейное положение клиентов влияет на факт погашения кредита в срок: одинокие клиенты и находящиеся в гражданском браке чаще нарушают сроки возврата кредитов, чем другие группы клиентов (на 3%)
* семьи с детьми нарушают сроки возвратов немного чаще, чем остальные группы клиентов (на 2%);
* количество детей влияет на факт возврата кредитов в срок, но не существенно (0,3%);
* наиболшее количество кредитных должников в группе с кредитами на операции с автомобилем (на 3%).
    </div>